### PR TITLE
FilterSamples IR

### PR DESCRIPTION
**Pull request by cseed**
Builds on: https://github.com/hail-is/hail/pull/2563.  Didn't stack (yet).

Had to remove the requiredness on gs since I can't produce it in the IR (minor).

Fixed two bugs along the way: wrong TypeInfo in ArrayMap and loadPrimitive wasn't loading the array address.

### CODE REVIEW

File: src/main/scala/is/hail/expr/Relational.scala [hunk 1]
```diff
@@ -1,7 +1,10 @@
 package is.hail.expr
 
+import java.io.PrintWriter
```
Comment (danking) [inline src/main/scala/is/hail/expr/Relational.scala]: stray import from debugging

**Comment by danking** [comment]
Test failures seem odd, they don't seem to use FilterSamples?

**Comment by cseed** [comment]
annotateVariantsVDS uses dropSamples (although technically it doesn't need to).

**Comment by danking** [comment]
In the second case it appears I might be treating Binary wrong?

**Comment by cseed** [comment]
That was it.  `loadPrimitive` also needed a call to `fundamentalType`.

File: src/main/scala/is/hail/utils/richUtils/RichCodeMemoryBuffer.scala [hunk 2]
```diff
@@ -83,18 +83,22 @@ class RichCodeRegion(val region: Code[Region]) extends AnyVal {
...
+      loadFloat
     case _: TFloat64 =>
-      this.loadDouble(_)
-    case _ =>
+      loadDouble
+    case _: TArray =>
+      loadAddress
+    case _: TBinary =>
+      loadAddress
+    case _: TStruct =>
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichCodeMemoryBuffer.scala]: I find catch all code like the old `case _ => off => off` to be somewhat dangerous (esp. if it is supposed to handle a few cases.  By enumerating them, (1) you have to think through the cases, (2) you check that verifies you did, and (3) by writing it out explicitly you make potentially make implicit errors in your assumptions explicit.)

Comment (danking) [reply src/main/scala/is/hail/utils/richUtils/RichCodeMemoryBuffer.scala]: 👍

**Pull request merged**